auth/standard_role_manager.cc

```diff
@@ -701,4 +703,67 @@ future<> standard_role_manager::remove_attribute(std::string_view role_name, std
                 {sstring(role_name), sstring(attribute_name)});
     }
 }
+
+future<std::vector<cql3::description>> standard_role_manager::describe_role_grants() {
+    std::vector<cql3::description> result{};
+
+    const auto grants = co_await query_all_directly_granted();
+    result.reserve(grants.size());
+
+    for (const auto& [grantee_role, granted_role] : grants) {
+        const auto formatted_grantee = cql3::util::maybe_quote(grantee_role);
+        const auto formatted_granted = cql3::util::maybe_quote(granted_role);
+
+        result.push_back(cql3::description {
+            // Role grants do not belong to any keyspace.
+            .keyspace = std::nullopt,
+            .type = "grant_role",
+            .name = granted_role,
+            .create_statement = seastar::format("GRANT {} TO {};", formatted_granted, formatted_grantee)
+        });
+
+        co_await coroutine::maybe_yield();
+    }
+
+    std::ranges::sort(result, std::less<>{}, [] (const cql3::description& desc) noexcept {
+        return std::make_tuple(std::ref(desc.name), std::ref(*desc.create_statement));
+    });
+
+    co_return result;
+}
+
+future<std::vector<cql3::description>> standard_role_manager::describe_attached_service_levels() {
```

Comment:
I wonder whether it makes sense to describe service level attachments to roles in auth? Isn't it more of a service-level specific concept? Does the mechanism used to attach service levels depend on the role manager? cc: @Jadw1 @nuivall

Reply:
I was thinking about the same, but I'm still not sure. Service levels are currently the only type of attributes that can be attached to a role, but there might be more in the future. Those attributes are stored in a system table related to auth, which would most likely make it more convenient to keep the code here.

Another, perhaps more down-to-earth, argument is that `service_level_controller` depends on the role manager to figure out what service level a given role uses. If we moved the code to `service/`, we'd avoid a few lines of code in `auth/`, but we'd need to query the role manager anyway. The control flow would still go through there, but instead of doing it directly via `auth`, we'd go `service -> auth -> service`.

Reply:
The way I see it is that service levels is a higher-level module than auth. Auth isn't aware of the service levels, it only exposes a system of generic attributes which are just string -> string maps. Service levels use it to store information and knows how to interpret it. Introducing the `describe_attached_service_levels` method to the`role_manager` interface would violate this principle.

Besides, it doesn't make sense for this method to be virtual in `role_manager`. It only uses other virtual methods of the `role_manager` interface, its implementation would be the same for the LDAP role manager in enterprise, so it could as well be non-virtual. However, I think we should go a step further and move it to the service levels code as it just doesn't belong in auth.

Reply:
I agree SL API is in service_level_controller (`state.get_service_level_controller()`)

It's broken in for instance in attach role statement `co_await auth::set_attribute(as, _role_name, "service_level", _service_level, mc);`

but this should be fixed and SL should take Auth service as dependency (perhaps fix not in this PR though)